File: c/kit/task2.c
/**
 * Derek Hessinger
 * Prof. Ying Li
 * CS 333
 * 12/6/24
 * 
 * This file performs a pixel wise operation to a given ppg file, taking in
 * arguments for the file name and the number of threads to use
 * 
 * gcc -o task2 -I. task2.c ppmIO.c my_timing.c -lm
 * ./task2 <file name> <num of threads>
 * 
 */

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include "ppmIO.h"
#include "my_timing.h"

#define ITERATIONS 100  // Number of times to process image for accurate timing

// Structure to pass data to threads
typedef struct {
    Pixel* image;
    int start_idx;
    int end_idx;
} ThreadData;

// Thread function to process image chunk
void* process_chunk(void* arg) {
    ThreadData* data = (ThreadData*)arg;
    
    // Process the chunk multiple times
    for(int iter = 0; iter < ITERATIONS; iter++) {
        for(int i = data->start_idx; i < data->end_idx; i++) {
            data->image[i].r = data->image[i].r > 128 ? (220 + data->image[i].r)/2 : (30 + data->image[i].r)/2;
            data->image[i].g = data->image[i].g > 128 ? (220 + data->image[i].g)/2 : (30 + data->image[i].g)/2;
            data->image[i].b = data->image[i].b > 128 ? (220 + data->image[i].b)/2 : (30 + data->image[i].b)/2;
        }
    }
    
    pthread_exit(NULL);
}

int main(int argc, char *argv[]) {
    Pixel *src;
    int rows, cols, colors;
    int num_threads;
    double start_time, end_time;
    
    // Check usage
    if(argc < 3) {
        printf("Usage: %s <image filename> <number of threads (1,2,4)>\n", argv[0]);
        exit(-1);
    }
    
    // Get number of threads
    num_threads = atoi(argv[2]);
    if(num_threads != 1 && num_threads != 2 && num_threads != 4) {
        printf("Number of threads must be 1, 2, or 4\n");
        exit(-1);
    }
    
    // Read image
    src = ppm_read(&rows, &cols, &colors, argv[1]);
    if(!src) {
        printf("Unable to read file %s\n", argv[1]);
        exit(-1);
    }
    
    // Calculate total pixels
    int total_pixels = rows * cols;
    
    // Create thread arrays
    pthread_t threads[4];
    ThreadData thread_data[4];
    
    // Start timing
    start_time = get_time_sec();
    
    // Create threads to process image chunks
    for(int i = 0; i < num_threads; i++) {
        thread_data[i].image = src;
        thread_data[i].start_idx = (total_pixels / num_threads) * i;
        thread_data[i].end_idx = (i == num_threads - 1) ? 
                                total_pixels : 
                                (total_pixels / num_threads) * (i + 1);
        
        int rc = pthread_create(&threads[i], NULL, process_chunk, (void*)&thread_data[i]);
        if(rc) {
            printf("Error creating thread %d\n", i);
            exit(-1);
        }
    }
    
    // Wait for all threads to complete
    for(int i = 0; i < num_threads; i++) {
        pthread_join(threads[i], NULL);
    }
    
    // End timing
    end_time = get_time_sec();
    
    // Calculate and print timing results
    double total_time = end_time - start_time;
    printf("Image processing completed:\n");
    printf("Number of threads: %d\n", num_threads);
    printf("Number of iterations: %d\n", ITERATIONS);
    printf("Total time: %.3f seconds\n", total_time);
    printf("Time per iteration: %.3f seconds\n", total_time / ITERATIONS);
    printf("Image dimensions: %d x %d (%d pixels)\n", rows, cols, total_pixels);
    
    // Write output image
    ppm_write(src, rows, cols, colors, "updated_img.ppm");
    
    // Clean up
    free(src);
    
    return 0;
}

File: ocaml/task2.ml
(**
 * Derek Hessinger
 * Prof. Ying Li
 * CS 333
 * 12/6/24
 * 
 * This file performs a pixel wise operation to a given ppg file, taking in
 * arguments for the file name and the number of threads to use
 * 
 *  ocamlopt -o task2 task2.ml
 * ./task2 <file name> <num of threads>
*)

open Printf
open Domain

(*create pixel type*)
type pixel = {
  mutable r : int;
  mutable g : int;
  mutable b : int;
}

(*create image type*)
type image = {
  width : int;
  height : int;
  max_color : int;
  pixels : pixel array;
}

(*num iterations*)
let iterations = 100

(* get current time in seconds *)
let get_time () = Sys.time()

(* process a single pixel *)
let process_pixel px =
  px.r <- if px.r > 128 then (200 + px.r) / 2 else (30 + px.r) / 2;
  px.g <- if px.g > 128 then (200 + px.g) / 2 else (30 + px.g) / 2;
  px.b <- if px.b > 128 then (200 + px.b) / 2 else (30 + px.b) / 2

(* Process a chunk of the image *)
let process_chunk start_idx end_idx pixels =
  for _iter = 1 to iterations do
    for i = start_idx to end_idx - 1 do
      process_pixel pixels.(i)
    done
  done

(* read in ppm image *)
let read_ppm filename =
  
  let ic = open_in_bin filename in
  let line = input_line ic in
  if line <> "P6" then failwith "Not a PPM file";
  let rec read_header () =
    let line = input_line ic in
    if line.[0] = '#' then read_header ()
    else line
  in
  let dims = read_header () in
  let [width; height] = List.map int_of_string (String.split_on_char ' ' dims) in
  let max_color = int_of_string (input_line ic) in
  let pixels = Array.init (width * height) (fun _ ->
    let r = input_byte ic in
    let g = input_byte ic in
    let b = input_byte ic in
    {r; g; b}
  ) in
  close_in ic;
  {width; height; max_color; pixels}

(* Write PPM image - this would need to interface with your ppmIO *)
let write_ppm img filename =
  let oc = open_out_bin filename in
  fprintf oc "P6\n%d %d\n%d\n" img.width img.height img.max_color;
  Array.iter (fun px ->
    output_byte oc px.r;
    output_byte oc px.g;
    output_byte oc px.b;
  ) img.pixels;
  close_out oc

(* Main parallel processing function *)
let process_image_parallel num_threads img =
  let total_pixels = Array.length img.pixels in
  let chunk_size = total_pixels / num_threads in
  
  let start_time = get_time () in
  
  (* Create and run domains *)
  let domains = Array.init num_threads (fun i ->
    let start_idx = i * chunk_size in
    let end_idx = if i = num_threads - 1 then total_pixels 
                  else (i + 1) * chunk_size in
    Domain.spawn (fun () -> 
      process_chunk start_idx end_idx img.pixels
    )
  ) in
  
  (* Wait for all domains to complete *)
  Array.iter Domain.join domains;
  
  let end_time = get_time () in
  let total_time = end_time -. start_time in
  
  (* Print timing results *)
  printf "Image processing completed:\n";
  printf "Number of threads: %d\n" num_threads;
  printf "Number of iterations: %d\n" iterations;
  printf "Total time: %.3f seconds\n" total_time;
  printf "Time per iteration: %.3f seconds\n" (total_time /. float_of_int iterations);
  printf "Image dimensions: %d x %d (%d pixels)\n" 
    img.width img.height total_pixels

let () =
  if Array.length Sys.argv < 3 then begin
    printf "Usage: %s <image filename> <number of threads (1,2,4)>\n" Sys.argv.(0);
    exit 1
  end;
  
  let num_threads = int_of_string Sys.argv.(2) in
  if not (List.mem num_threads [1; 2; 4]) then begin
    printf "Number of threads must be 1, 2, or 4\n";
    exit 1
  end;
  
  try
    let img = read_ppm Sys.argv.(1) in
    process_image_parallel num_threads img;
    write_ppm img "updated_img.ppm"
  with 
  | Sys_error msg -> 
      printf "Error reading/writing file: %s\n" msg;
      exit 1
  | Failure msg -> 
      printf "Error processing image: %s\n" msg;
      exit 1